docs/proposals/004_carbon_brd.md

```diff
@@ -0,0 +1,109 @@
+
+**Standardized Mounting Interface**
+
+- **Mount API:** The **Carbon** platform must provide a consistent mounting interface for all extension types (e.g., `mount(container, props)`). This interface should allow colocated and remote extensions to be mounted dynamically within a specific DOM container while passing the necessary props and configuration details from the manifest.
+- **Lazy Loading:** The **Carbon** system should support lazy loading of extensions, allowing resources to be fetched and initialized only when needed, optimizing performance and reducing the initial load time.
+
+---
+
+### 3. Communication
+
+A secure and standardized communication channel is essential for interactions between the host app and its extensions, particularly iframe-based and remote apps in **Carbon**.
+
+#### Requirements:
+
+- **Client API for Communication:** For iframe-based extensions, a `clientApi` or equivalent interface should be established. This API should allow two-way communication between the host app and the iframe, enabling the exchange of data, events, and state while maintaining security.
+- **Message Passing Mechanism:** Implement a secure, standardized message-passing system (e.g., postMessage API) that works across all extension types—whether colocated, remote, or iframes. This system should enable event-driven communication and data exchange between the host and extensions, ensuring that only authorized messages are processed.
+
+- **Event Broadcasting:** The **Carbon** platform should enable extensions to broadcast and listen to events, allowing for inter-app communication when necessary, but such interactions must be governed by strict security and isolation protocols.
+
+---
+
+### 4. Security Considerations
+
+Security is a critical aspect of loading external extensions in **Carbon**, particularly when dealing with remote or iframe-based apps.
+
+#### Requirements:
+
+- **Sandboxing for Iframes:** All iframe-based extensions should be sandboxed, enforcing restrictions like `allow-scripts`, `allow-same-origin`, and other relevant flags to limit potential security risks (e.g., cross-site scripting, data leakage).
+- **CORS and Content Security Policy (CSP):** For remote extensions, the platform must enforce proper CORS headers and CSP to ensure that only authorized domains are allowed to load and interact with the platform.
+
+- **Authentication and Authorization:** Extensions must adhere to the **Carbon** platform's authentication and authorization mechanisms, ensuring that only authenticated users with appropriate permissions can load or interact with specific extensions.
+
+- **Dependency Security:** The manifest file or configuration source must define any external dependencies required by the extension. These dependencies should be validated for security vulnerabilities before being loaded into the platform.
```

Comment:
validating or checking security vulnerabilities of external dependencies might not be possible for **remote extensions** because if I understood correctly they would already come as final executable/importable bundle meaning all of the external dependencies would already be packaged.

Reply:
> validating or checking security vulnerabilities of external dependencies might not be possible for **remote extensions** because if I understood correctly they would already come as final executable/importable bundles meaning all of the external dependencies would already be packaged inside those bundles.

I believe the intention is that no extensions will be loaded or rendered until the user is authenticated in the shell app. Once authenticated, any necessary properties can then be passed to the extensions as needed.

Reply:
I appreciate the discussion here; it’s exactly what this space is meant for! 👍 My original thought was that bundling the extensions at build time would allow us to run checks on the interface and code quality using tools like Trivy. However, I understand that this approach wouldn’t be feasible for iframe-based extensions. For those, we would likely need to rely on a manual process for validation.

Reply:
exactly for third party extension running in an iframe or coming as importable bundles we can't verify their dependencies so it is up to the user to do the due diligence before integrating them to the system. From our side we can only render them in a sandboxed/secure environment so they don't mess up with the shell or other loaded extensions.